File: cli/internal_testsuites/golang/test_helpers/enclave_setup.go
package test_helpers

import (
	"context"
	"fmt"
	"github.com/kurtosis-tech/kurtosis-core-api-lib/api/golang/lib/enclaves"
	"github.com/kurtosis-tech/kurtosis-engine-api-lib/api/golang/lib/kurtosis_context"
	"github.com/sirupsen/logrus"
	"github.com/stretchr/testify/require"
	"testing"
	"time"
)

const (
	testsuiteNameEnclaveIDFragment = "go-testsuite"

	millisInNanos = 1000
)

func CreateEnclave(t *testing.T, ctx context.Context, testName string, isPartitioningEnabled bool) (resultEnclaveCtx *enclaves.EnclaveContext, resultStopEnclaveFunc func(), resultDestroyEnclaveFunc func() error, resultErr error) {
	kurtosisCtx, err := kurtosis_context.NewKurtosisContextFromLocalEngine()
	require.NoError(t, err, "An error occurred connecting to the Kurtosis engine for running test '%v'", testName)
	enclaveId := enclaves.EnclaveID(fmt.Sprintf(
		"%v.%v.%v",
		testsuiteNameEnclaveIDFragment,
		testName,
		time.Now().UnixNano() / millisInNanos,
	))
	enclaveCtx, err := kurtosisCtx.CreateEnclave(ctx, enclaveId, isPartitioningEnabled)
	require.NoError(t, err, "An error occurred creating enclave '%v'", enclaveId)
	stopEnclaveFunc := func() {
		if err := kurtosisCtx.StopEnclave(ctx, enclaveId); err != nil {
			logrus.Errorf("An error occurred stopping enclave '%v' that we created for this test:\n%v", enclaveId, err)
			logrus.Errorf("ACTION REQUIRED: You'll need to stop enclave '%v' manually!!!!", enclaveId)
		}
	}
	destroyEnclaveFunc := func() error {
		if err := kurtosisCtx.DestroyEnclave(ctx, enclaveId); err != nil {
			logrus.Errorf("An error occurred destroying enclave '%v' that we created for this test:\n%v", enclaveId, err)
			logrus.Errorf("ACTION REQUIRED: You'll need to destroy enclave '%v' manually!!!!", enclaveId)
			return err
		}
		return nil
	}

	return enclaveCtx, stopEnclaveFunc, destroyEnclaveFunc, nil
}

File: cli/internal_testsuites/typeScript/src/test_helpers/enclave_setup.ts
import { EnclaveContext, EnclaveID } from "kurtosis-core-api-lib"
import { KurtosisContext,  } from "kurtosis-engine-api-lib"
import {Result, err, ok} from "neverthrow"
import log from "loglevel";

const TEST_SUITE_NAME_ENCLAVE_ID_FRAGMENT = "typescript-engine-server-test";
const MILLISECONDS_IN_SECOND = 1000;

export async function createEnclave(testName:string, isPartitioningEnabled: boolean):
	Promise<Result<{ 
        enclaveContext: EnclaveContext, 
        stopEnclaveFunction: () => void
    }, Error>> {

	const newKurtosisContextResult = KurtosisContext.newKurtosisContextFromLocalEngine();
	if(newKurtosisContextResult.isErr()) {
		return err(new Error(`An error occurred connecting to the Kurtosis engine for running test ${testName}`))
	}
	const kurtosisContext = newKurtosisContextResult.value;
	
	const enclaveId:EnclaveID = `${TEST_SUITE_NAME_ENCLAVE_ID_FRAGMENT}_${testName}_${Math.round(Date.now()/MILLISECONDS_IN_SECOND)}`
	const createEnclaveResult = await kurtosisContext.createEnclave(enclaveId, isPartitioningEnabled);
	
	if(createEnclaveResult.isErr()) {
		return err(new Error(`An error occurred creating enclave ${enclaveId}`))
	}

	const enclaveContext = createEnclaveResult.value;

	const stopEnclaveFunction = async ():Promise<void> => {
		const stopEnclaveResult = await kurtosisContext.stopEnclave(enclaveId)
		if(stopEnclaveResult.isErr()) {
			log.error(`An error occurred stopping enclave ${enclaveId} that we created for this test: ${stopEnclaveResult.error.message}`)
			log.error(`ACTION REQUIRED: You'll need to stop enclave ${enclaveId} manually!!!!`)
		}
	}

	return ok({
		enclaveContext,
		stopEnclaveFunction
	})
}
